### PR TITLE
Refactoring of crac io api  and implementation of Crac/Json converter

### PR DESCRIPTION
**Pull request by TheMaskedTurtle**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Change the interface of importers/exporters services for Crac management. And gives the first implementation of this interface for Json format. 

**What is the current behavior?** *(You can also link to an open issue here)*
Not implemented interfaces.

**What is the new behavior (if this is a feature change)?**
New static classes in the API which will be based on the previous import and export interfaces. We can now use a importers and exporters with Json files.

**Does this PR introduce a breaking change?** *(What changes might users need to make in their application due to this PR?)*
It introduces a change in the way the module farao-crac-io-api would be used.

### CODE REVIEW

File: data/crac-file/crac-io-api/pom.xml [hunk 1]
```diff
@@ -15,11 +15,27 @@
...
         </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-all</artifactId>
+            <version>1.10.19</version>
```
Comment (pengbo-ruidian) [inline data/crac-file/crac-io-api/pom.xml]: version number be in the main pom.xml

File: data/crac-file/crac-io-api/src/test/java/com/farao_community/farao/data/crac_io_api/CracExporterMock.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package com.farao_community.farao.data.crac_io_api;
```
Comment (pengbo-ruidian) [inline data/crac-file/crac-io-api/src/test/java/com/farao_community/farao/data/crac_io_api/CracExporterMock.java]: copyrigh header?

File: data/crac-file/crac-io-api/src/test/java/com/farao_community/farao/data/crac_io_api/CracExportersImportersTest.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
+package com.farao_community.farao.data.crac_io_api;
```
Comment (pengbo-ruidian) [inline data/crac-file/crac-io-api/src/test/java/com/farao_community/farao/data/crac_io_api/CracExportersImportersTest.java]: header copyright

File: data/crac-file/crac-io-json/pom.xml [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            <groupId>com.powsybl</groupId>
+            <artifactId>powsybl-commons</artifactId>
+            <version>${powsybl.core.version}</version>
+            <type>test-jar</type>
+            <scope>test</scope>
+        </dependency>
+        <dependency> <!--for CracSchemaGeneration-->
+            <groupId>com.fasterxml.jackson.module</groupId>
+            <artifactId>jackson-module-jsonSchema</artifactId>
+            <version>2.3.0</version>
```
Comment (pengbo-ruidian) [inline data/crac-file/crac-io-json/pom.xml]: move version number?

File: data/crac-file/crac-io-json/pom.xml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.datatype</groupId>
+            <artifactId>jackson-datatype-jdk8</artifactId>
+            <version>2.9.6</version>
```
Comment (pengbo-ruidian) [inline data/crac-file/crac-io-json/pom.xml]: bis

File: data/crac-file/crac-io-json/src/main/java/com/farao_community/farao/data/crac_io_json/JsonImport.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    @Override
+    public boolean exists(String fileName, InputStream inputStream) {
+        return validCracFile(fileName, inputStream);
+    }
+
+    private boolean validCracFile(String fileName, InputStream inputStream) {
+        JSONObject jsonSubject = new JSONObject(
+            new JSONTokener(inputStream));
+        SCHEMA_JSON.validate(jsonSubject);
+        return fileName.split("\\.")[1].equals(JSON_EXTENSION);
```
Comment (pengbo-ruidian) [inline data/crac-file/crac-io-json/src/main/java/com/farao_community/farao/data/crac_io_json/JsonImport.java]: I would use something in a lib like FilenameUtils.getExtension; or something like fileName.lastIndexOf('.');
ref https://stackoverflow.com/questions/3571223/how-do-i-get-the-file-extension-of-a-file-in-java

File: data/crac-file/crac-io-json/src/test/resources/file.json [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+        "@c" : ".Instant",
+        "duration" : 200.0
+      }
+    }
+  }, {
+    "@c" : "com.farao_community.farao.data.crac_impl.SimpleCnec",
+    "id" : "idCnec2",
+    "name" : "Cnec 2",
+    "criticalNetworkElement" : {
+      "@c" : ".NetworkElement",
```
Comment (baptsegu) [inline data/crac-file/crac-io-json/src/test/resources/file.json]: Why do we need to precise that this is a NetworkElement, can it be something else ?

File: data/crac-file/crac-io-json/src/test/resources/file.json [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+      "min" : 100.0,
+      "max" : 1000.0
+    }, {
+      "@c" : "com.farao_community.farao.data.crac_impl.range_domain.RelativeFixedRange",
+      "min" : 1.0,
+      "max" : 10.0
+    } ],
+    "applicableRangeActions" : [ {
+      "@c" : "com.farao_community.farao.data.crac_impl.remedial_action.range_action.PstRange",
+      "networkElement" : {
```
Comment (baptsegu) [inline data/crac-file/crac-io-json/src/test/resources/file.json]: same remark for network Elements within RA

File: pom.xml [hunk 9]
```diff
@@ -253,19 +253,6 @@
...
-                        <version>${maven.javadoc.version}</version>
-                        <executions>
-                            <execution>
-                                <phase>package</phase>
-                                <goals>
-                                    <goal>jar</goal>
-                                </goals>
-                            </execution>
-                        </executions>
-                    </plugin>
```
Comment (baptsegu) [inline pom.xml]: Why should we remove this ??

Comment (terriervik) [reply pom.xml]: I don't know.

File: data/crac-file/crac-io-api/pom.xml [hunk 10]
```diff
@@ -25,6 +25,16 @@
             <artifactId>farao-crac-api</artifactId>
             <version>${project.version}</version>
         </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
```
[the remark this answers is not in the document]
Comment (terriervik) [reply data/crac-file/crac-io-api/pom.xml]: ok

File: data/crac-file/crac-io-json/src/test/java/com/farao_community/farao/data/crac_io_json/JsonImportExportTest.java [hunk 11]
```diff
@@ -195,4 +200,28 @@ public void testExportCrac() throws IOException {
         os.flush();
         os.close();
     }
```
[the remark this answers is not in the document]
Comment (terriervik) [reply data/crac-file/crac-io-json/src/test/java/com/farao_community/farao/data/crac_io_json/JsonImportExportTest.java]: ok

File: data/crac-file/crac-io-api/src/main/java/com/farao_community/farao/data/crac_io_api/CracImporters.java [hunk 12]
```diff
@@ -30,28 +30,46 @@ private CracImporters() {
...
 
-    public static Crac importCrac(InputStream inputStream) {
-        CracImporter importer = findImporter(inputStream);
-        if (importer == null) {
-            throw new FaraoException("No importer found for this file");
+    public static Crac importCrac(String fileName, InputStream inputStream) {
+        try {
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            org.apache.commons.io.IOUtils.copy(inputStream, baos);
+            byte[] bytes = baos.toByteArray();
```
[the remark this answers is not in the document]
Comment (terriervik) [reply data/crac-file/crac-io-api/src/main/java/com/farao_community/farao/data/crac_io_api/CracImporters.java]: ok

File: data/crac-file/crac-io-json/src/test/resources/file.json (hunk 7 above)
Comment (terriervik) [reply data/crac-file/crac-io-json/src/test/resources/file.json]: no

File: data/crac-file/crac-io-json/src/test/resources/file.json (hunk 8 above)
Comment (terriervik) [reply data/crac-file/crac-io-json/src/test/resources/file.json]: same

**Pull request merged**